sale_order_product_recommendation/wizards/sale_order_recommendation.py

```diff
@@ -6,6 +6,12 @@
 from odoo import api, fields, models
 
 
+def _onchange_method(record, field_list):
+    for field in field_list:
+        for onchange_method in record._onchange_methods[field]:
```

Comment:
@Yajo can you link here the piece of code which Odoo uses (you have done it in previous occasions) for using that one instead?

Reply:
It's not very easy. There's an addon for that, you can use it if you want: https://github.com/OCA/server-tools/blob/11.0/onchange_helper/models/base.py

Reply:
No, that's not the one I'm referring, and I prefer to avoid that module, as it targets to get the dictionary vals, not what we want. I mean the one in Odoo core.

Reply:
Does this depend on on_change_helper?

Reply:
No, it's a "manual way" of doing something similar

Reply:
do you mean the core function ?https://github.com/odoo/odoo/blob/d6ad22227cc7b3f36ac9914cce4f8a1f0a0f36a2/odoo/models.py#L577-L604

or this part?
```
        if onchange in ("1", "true"):
            for method in self._onchange_methods.get(field_name, ()):
                method_res = method(self)
                process(method_res)
            return
```

Reply:
@tbaden yes, it seems to be that. I saw it in the past, but didn't remember where. Hopefully with odoo/odoo#32784, we will avoid this need.

Reply:
The method that the ORM calls is [`onchange()`](https://github.com/odoo/odoo/blob/d6ad22227cc7b3f36ac9914cce4f8a1f0a0f36a2/odoo/models.py#L5052-L5092).

In v12 we can use [the `Form` class from the testing suite](https://www.odoo.com/documentation/12.0/reference/testing.html#odoo.tests.common.Form), although using it for non-tests could be considered abusing the feature.

AFAICS the sanest and best supported way to do this is with the `onchange_helper` module, which is a community-supported helper done precisely for this kind of usages.

Reply:
for future reference this part is also in the models.py
https://github.com/odoo/odoo/blob/d6ad22227cc7b3f36ac9914cce4f8a1f0a0f36a2/odoo/models.py#L5019-L5023

Reply:
The more that I look into this issue, the more I get convinced you need to use the `onchange_helper` addon or the `onchange()` method. Let me explain:

- [As you can see](https://github.com/odoo/odoo/blob/d6ad22227cc7b3f36ac9914cce4f8a1f0a0f36a2/odoo/models.py#L4997-L5001), it's legal for an onchange method to return a dictionary like `{'value': {...}}` instead of modifying `self`. This is legal and required sometimes because it's also legal to return other keys like `domain` or `warning` in that dict. You're not supporting this method here.
- [Onchanges can expect having a `__onchange: True` key in their context, or having a `self._origin` attribute pointing to the original record. Also they are expected to be executed in a `newId` record. This is also done in a different context, with different cache. Not supporting this can lead to unexpected results](https://github.com/odoo/odoo/blob/d6ad22227cc7b3f36ac9914cce4f8a1f0a0f36a2/odoo/models.py#L5158-L5161).
- The `newId` record is created from the values that the `onchange` method gets, not from the values that the current record has (in upstream odoo; [`onchange_helper` handles this](https://github.com/OCA/server-tools/blob/e0d943466a4e545bed01b48a126094a703ae0628/onchange_helper/models/base.py#L56)).

Summarizing, it has many corner cases because it's supposed to be used from the web client only, not from the backend; and it's designed as such. To bridge that gap, the safest and sanest way to work seems to be `onchange_helper`, which is not perfect but handles all you need.